File: symfony/src/Service/Marketplace/TradeTransactionService.php
<?php

namespace App\Service\Marketplace;

use App\Entity\Marketplace\TradeTransaction;
use App\Repository\Marketplace\TradeTransactionRepository;

class TradeTransactionService
{
    public function __construct(
        private TradeTransactionRepository $repository,
    ) {}

    public function create(array $data): TradeTransaction
    {
        throw new \LogicException('Not implemented');
    }

    public function update(TradeTransaction $entity, array $data): TradeTransaction
    {
        throw new \LogicException('Not implemented');
    }

    public function complete(int $id): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('TradeTransaction not found: ' . $id);
        $entity->complete();
        $this->repository->save($entity, flush: true);
    }

    public function refund(int $id): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('TradeTransaction not found: ' . $id);
        $entity->refund();
        $this->repository->save($entity, flush: true);
    }

    public function openDispute(int $id, $reason): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('TradeTransaction not found: ' . $id);
        $entity->openDispute($reason);
        $this->repository->save($entity, flush: true);
    }

    public function sellerNet(int $id): mixed
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('TradeTransaction not found: ' . $id);
        $result = $entity->sellerNet();
        $this->repository->save($entity, flush: true);
        return $result;
    }
}

File: rails/app/services/marketplace/trade_transaction_service.rb
module Marketplace
  class TradeTransactionService
    def initialize(repository = TradeTransaction)
      @repository = repository
    end

    def create(attributes)
      raise NotImplementedError, "#{self.class}#create not implemented"
    end

    def update(entity, attributes)
      raise NotImplementedError, "#{self.class}#update not implemented"
    end

    def complete(id)
      instance = TradeTransaction.find(id)
      instance.complete()
      instance.save!
    end

    def refund(id)
      instance = TradeTransaction.find(id)
      instance.refund()
      instance.save!
    end

    def open_dispute(id, reason)
      instance = TradeTransaction.find(id)
      instance.open_dispute(reason)
      instance.save!
    end

    def seller_net(id)
      instance = TradeTransaction.find(id)
      result = instance.seller_net()
      instance.save!
      result
    end
  end
end
